cmake/ImportBSON.cmake

```diff
@@ -0,0 +1,170 @@
+            "module is not available. Upgrade your CMake version, or provide a "
+            "MONGOCRYPT_MONGOC_DIR path to a mongo-c-driver directory (This is required "
+            "for libmongocrypt to find a libbson to use and link against).")
+   endif ()
+   include (FetchMongoC)
+   # The FetchMongoC module defines a MONGOCRYPT_MONGOC_DIR for use to use
+endif ()
+
+message (STATUS "Using [${MONGOCRYPT_MONGOC_DIR}] as a sub-project for libbson")
+
+function (_import_bson_add_subdir)
+   # Disable AWS_AUTH, to prevent it from building the kms-message symbols, which we build ourselves
+   set (ENABLE_MONGODB_AWS_AUTH OFF CACHE BOOL "Disable kms-message content in mongoc for libmongocrypt" FORCE)
+   # Disable install() for the libbson static library. We'll do it ourselves
+   set (ENABLE_STATIC BUILD_ONLY)
+   # Disable over-alignment of bson types
+   set (ENABLE_EXTRA_ALIGNMENT ${_extra_alignment_default} CACHE BOOL "Toggle extra alignment of bson_t")
+   # Add the subdirectory as a project. EXCLUDE_FROM_ALL to inhibit building and installing of components unless requested
+   add_subdirectory ("${MONGOCRYPT_MONGOC_DIR}" _mongo-c-driver EXCLUDE_FROM_ALL)
+endfunction ()
+
+# Do the add_subdirectory() in a function to isolate variable scope
+_import_bson_add_subdir ()
+
+# Define interface targets to be used to control the libbson used at both build and import time.
+# Refer to mongocrypt-config.cmake to see how these targets are used by consumers
+add_library (_mongocrypt-libbson_for_static INTERFACE)
+add_library (_mongocrypt-libbson_for_shared INTERFACE)
+add_library (_mongocrypt::libbson_for_static ALIAS _mongocrypt-libbson_for_static)
+add_library (_mongocrypt::libbson_for_shared ALIAS _mongocrypt-libbson_for_shared)
+install (
+   TARGETS _mongocrypt-libbson_for_static _mongocrypt-libbson_for_shared
+   EXPORT mongocrypt_targets
+)
+
+# Link to the requested libbson, only exporting that usage for the local build tree.
+# The mongocrypt-config file will later add the appropriate link library for downstream
+# users during find_package()
+if (USE_SHARED_LIBBSON)
+   target_link_libraries (_mongocrypt-libbson_for_shared INTERFACE $<BUILD_INTERFACE:bson_shared>)
+else ()
+   target_link_libraries (_mongocrypt-libbson_for_shared INTERFACE $<BUILD_INTERFACE:bson_static>)
+endif ()
+# libbson_for_static always links to the static libbson:
+target_link_libraries (_mongocrypt-libbson_for_static INTERFACE $<BUILD_INTERFACE:bson_static>)
+
+# And an alias to the mongoc target for use in some test cases
+add_library (_mongocrypt::mongoc ALIAS mongoc_static)
+# Workaround: Embedded mongoc_static does not set its INCLUDE_DIRECTORIES for user targets
+target_include_directories (mongoc_static
+   PUBLIC
+      "$<BUILD_INTERFACE:${MONGOCRYPT_MONGOC_DIR}/src/libmongoc/src>"
+      "$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/_mongo-c-driver/src/libmongoc/src/mongoc>"
+   )
```

Comment:
I see no benefit to provide end-users (not package-maintainers) the option to import the system's libbson and libmongoc as a supported configuration option. If a user wishes to provide a tweaked version of either library, they can do so by setting `MONGOCRYPT_MONGOC_DIR` to a path to a local copy of the mongo-c-driver. If they wish to link their final application against the system's libbson (static or shared), this will still "just work" even if we `add_subdirectory()` a copy of mongo-c-driver.

Reply:
> I see no benefit to provide end-users (not package-maintainers) the option to import the system's libbson and libmongoc as a supported configuration option.

I was not distinguishing between "end-users" and "package-maintainers", as these configuration variables are all public (regardless of documentation). However, it seems you have effectively managed to (partially) implemented the feature I was requesting in https://github.com/mongodb/libmongocrypt/pull/365/commits/b91cfd77713849dc1da9957352f857636283fe81 via the `MONGOCRYPT_MONGOC_DIR` variable instead. This obviates the need to use the `<library>_DIR` variables as I was suggesting.

The intent of these suggestions can now be interpreted as extending the behavior of `MONGOCRYPT_MONGOC_DIR` to include the libmongoc library being linked against by libmongocrypt online tests for consistency. Given latest state of PR (as of 5c6c431d), configuring libmongocrypt with `-DMONGOCRYPT_MONGOC_DIR=USE-SYSTEM -DUSE_SHARED_LIBBSON=ON -DENABLE_ONLINE_TESTS=ON` produces the following configuration error(s):
```
CMake Error at cmake/ImportBSON.cmake:170 (add_library):
  add_library cannot create ALIAS target "_mongocrypt::mongoc" because target
  "mongoc_static" does not already exist.
Call Stack (most recent call first):
  CMakeLists.txt:24 (include)
CMake Error at cmake/ImportBSON.cmake:172 (target_include_directories):
  Cannot specify include directories for target "mongoc_static" which is not
  built by this project.
Call Stack (most recent call first):
  CMakeLists.txt:24 (include)
CMake Error at cmake/ImportBSON.cmake:191 (install):
  install FILES given no DESTINATION!
Call Stack (most recent call first):
  CMakeLists.txt:24 (include)
```

Given these configuration variables, libmongocrypt's online tests should ideally link with the same libmongoc library being provided by `MONGOCRYPT_MONGOC_DIR` as already being done for libbson, rather than failing configuration. This includes `-DMONGOCRYPT_MONGOC_DIR=USE-SYSTEM -DUSE_SHARED_LIBBSON=OFF` (if we wish to support statically linking with system libbson/libmongoc libraries), which currently fails during configuration.

Reply:
My personal vote it to leave this particular configuration (wherein we use online-tests as well as `"USE-SYSTEM"`) unsupported, possibly issuing a warning to that effect, or possibly another guard around `"USE-SYSTEM"` to check that the online tests are disabled (as the only intended use case thereof is for package maintainers that want to dynamic link with the distribution's libraries).

Reply:
I can support that.  In a package maintenance context, I think that "fully offline" is the expected default.

Reply:
>  issuing a warning to that effect, or possibly another guard around "USE-SYSTEM" to check that the online tests are disabled
So long as it is made clear that the given configurations are not supported, SGTM.